src/benchmark.cc

```diff
@@ -198,140 +198,225 @@ struct RunResults {
-      if (multiplier <= 1.0) multiplier = 2.0;
-      double next_iters = std::max(multiplier * iters, iters + 1.0);
-      if (next_iters > kMaxIterations) {
-        next_iters = kMaxIterations;
+ public:
+  BenchmarkRunner(const benchmark::internal::Benchmark::Instance& b_,
+                  std::vector<BenchmarkReporter::Run>* complexity_reports_)
+      : b(b_),
+        complexity_reports(*complexity_reports_),
+        min_time(!IsZero(b.min_time) ? b.min_time : FLAGS_benchmark_min_time),
+        repeats(b.repetitions != 0 ? b.repetitions
+                                   : FLAGS_benchmark_repetitions),
+        has_explicit_iteration_count(b.iterations != 0),
+        pool(b.threads - 1),
+        iters(has_explicit_iteration_count ? b.iterations : 1) {
+    if (repeats != 1) {
+      run_results.display_report_aggregates_only =
+          (FLAGS_benchmark_report_aggregates_only ||
+           FLAGS_benchmark_display_aggregates_only);
+      run_results.file_report_aggregates_only =
+          FLAGS_benchmark_report_aggregates_only;
+      if (b.aggregation_report_mode != internal::ARM_Unspecified) {
+        run_results.display_report_aggregates_only =
+            (b.aggregation_report_mode &
+             internal::ARM_DisplayReportAggregatesOnly);
+        run_results.file_report_aggregates_only =
+            (b.aggregation_report_mode &
+             internal::ARM_FileReportAggregatesOnly);
       }
-      VLOG(3) << "Next iters: " << next_iters << ", " << multiplier << "\n";
-      iters = static_cast<int>(next_iters + 0.5);
     }
-  }
 
-  // Calculate additional statistics
-  run_results.aggregates_only = ComputeStats(run_results.non_aggregates);
+    for (int repetition_num = 0; repetition_num < repeats; repetition_num++) {
+      const bool not_in_the_first_repetition = repetition_num != 0;
+      doOneRepetition(not_in_the_first_repetition);
+    }
+
+    // Calculate additional statistics
+    run_results.aggregates_only = ComputeStats(run_results.non_aggregates);
 
-  // Maybe calculate complexity report
-  if ((b.complexity != oNone) && b.last_benchmark_instance) {
-    auto additional_run_stats = ComputeBigO(*complexity_reports);
-    run_results.aggregates_only.insert(run_results.aggregates_only.end(),
-                                       additional_run_stats.begin(),
-                                       additional_run_stats.end());
-    complexity_reports->clear();
+    // Maybe calculate complexity report
+    if ((b.complexity != oNone) && b.last_benchmark_instance) {
+      auto additional_run_stats = ComputeBigO(complexity_reports);
+      run_results.aggregates_only.insert(run_results.aggregates_only.end(),
+                                         additional_run_stats.begin(),
+                                         additional_run_stats.end());
+      complexity_reports.clear();
+    }
   }
 
-  return run_results;
+  RunResults getResults() { return run_results; }
```

Comment:
I'm torn between this, where you have a simple accessor but the constructor is complex and actually doing the work, and an alternative where the constructor merely initialises the state, and a method `Run` that runs the benchmark and returns `RunResults`.

Reply:
Yes, i don't like this, too.
The current approach (complex constructor) seems better,
because we really wouldn't want to re-run re-use
the same `BenchmarkRunner` instance more than once.